src/particles.cpp

```diff
@@ -30,623 +29,729 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-					event->add_particlespawner.collision_removal,
-					event->add_particlespawner.attached_id,
-					event->add_particlespawner.vertical,
-					texture,
-					event->add_particlespawner.id,
-					event->add_particlespawner.animation,
-					event->add_particlespawner.glow,
-					this);
-
-			/* delete allocated content of event */
-			delete event->add_particlespawner.minpos;
-			delete event->add_particlespawner.maxpos;
-			delete event->add_particlespawner.minvel;
-			delete event->add_particlespawner.maxvel;
-			delete event->add_particlespawner.minacc;
-			delete event->add_particlespawner.texture;
-			delete event->add_particlespawner.maxacc;
-
-			{
-				MutexAutoLock lock(m_spawner_list_lock);
-				m_particle_spawners.insert(
-						std::pair<u32, ParticleSpawner*>(
-								event->add_particlespawner.id,
-								toadd));
-			}
-			break;
+		scene::IParticleAffector *acceleration_affector =
+			new AccelerationAffector(*event->add_particlespawner.minacc,
+				*event->add_particlespawner.maxacc);
+		ps->addAffector(acceleration_affector);
+		acceleration_affector->drop();
+
+		if (event->add_particlespawner.collisiondetection) {
+			scene::IParticleAffector *collision_affector =
+				new CollisionAffector(m_env,
+					event->add_particlespawner.collision_removal);
+			ps->addAffector(collision_affector);
+			collision_affector->drop();
 		}
-		case CE_SPAWN_PARTICLE: {
-			video::ITexture *texture =
-				client->tsrc()->getTextureForMesh(*(event->spawn_particle.texture));
-
-			Particle *toadd = new Particle(client, player, m_env,
-					*event->spawn_particle.pos,
-					*event->spawn_particle.vel,
-					*event->spawn_particle.acc,
-					event->spawn_particle.expirationtime,
-					event->spawn_particle.size,
-					event->spawn_particle.collisiondetection,
-					event->spawn_particle.collision_removal,
-					event->spawn_particle.vertical,
-					texture,
-					v2f(0.0, 0.0),
-					v2f(1.0, 1.0),
-					event->spawn_particle.animation,
-					event->spawn_particle.glow);
-
-			addParticle(toadd);
-
-			delete event->spawn_particle.pos;
-			delete event->spawn_particle.vel;
-			delete event->spawn_particle.acc;
-			delete event->spawn_particle.texture;
-
-			break;
+
+		scene::IParticleAffector *lighting_affector =
+			new LightingAffector(m_env, client, 0);
+		ps->addAffector(lighting_affector);
+		lighting_affector->drop();
+
+		ps->setDebugDataVisible(irr::scene::EDS_BBOX);
+
+		// Delete allocated content of event
+		delete event->add_particlespawner.minpos;
+		delete event->add_particlespawner.maxpos;
+		delete event->add_particlespawner.minvel;
+		delete event->add_particlespawner.maxvel;
+		delete event->add_particlespawner.minacc;
+		delete event->add_particlespawner.maxacc;
+		delete event->add_particlespawner.texture;
+		break;
+	}
+	case CE_SPAWN_PARTICLE: {
+		scene::IParticleSystemSceneNode *ps =
+			m_smgr->addParticleSystemSceneNode(false,
+				RenderingEngine::get_scene_manager()->getRootSceneNode());
```

Comment:
I think particles with identical primary properties (texture) might share a particle system. That’s out of scope of this PR, though.